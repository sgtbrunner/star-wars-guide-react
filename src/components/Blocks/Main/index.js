import { connect } from 'react-redux';

import Main from './main.component';

const mapStateToProps = (state) => ({
  characters: state.characters.data,
});

export default connect(mapStateToProps)(Main);
