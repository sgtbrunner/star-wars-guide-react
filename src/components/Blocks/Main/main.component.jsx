import React, { useState } from 'react';
import PropTypes from 'prop-types';

import CardList from '../../Cardlist';
import Dialog from '../../Dialog';
import SearchBox from '../../SearchBox';
import { getStats, getFilms } from '../../../utils/functions.utils';
import './main.styles.css';

const Main = ({ characters }) => {
  const [searchField, setSearchField] = useState('');
  const [clickedCard, setClickedCard] = useState({});
  const [showModal, setShowModal] = useState(false);
  const [characterDetails, setCharacterDetails] = useState({ race: {}, planet: {}, movies: '' });
  const onSearchChange = (event) => {
    setSearchField(event.target.value);
  };

  const openModal = async (event) => {
    setShowModal(true);
    setClickedCard(characters[event.target.id]);
    const race = await getStats(characters[event.target.id].species);
    const planet = await getStats(characters[event.target.id].homeworld);
    const movies = await getFilms(characters[event.target.id].films);
    setCharacterDetails({
      race,
      planet,
      movies,
    });
  };

  const onCloseClick = () => {
    setCharacterDetails({
      race: {},
      planet: {},
      movies: '',
    });
    setClickedCard({});
    setShowModal(false);
  };

  const filteredCharacters = characters.filter((character) =>
    character.name.toLowerCase().includes(searchField.toLowerCase())
  );

  return (
    <div id="main">
      <SearchBox searchChange={onSearchChange} />
      <CardList characters={filteredCharacters} openModal={openModal} />
      ''
      {showModal && (
        <Dialog
          character={clickedCard}
          race={characterDetails.race}
          planet={characterDetails.planet}
          movies={characterDetails.movies}
          onCloseClick={onCloseClick}
        />
      )}
    </div>
  );
};

Main.propTypes = {
  characters: PropTypes.array,
};

Main.defaultProps = {
  characters: [],
};

export default Main;
