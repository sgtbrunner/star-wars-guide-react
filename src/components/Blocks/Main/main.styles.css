#main {
  background-image: url('../../../assets/img/bg.jpg');
  padding: 15px 0;
}
