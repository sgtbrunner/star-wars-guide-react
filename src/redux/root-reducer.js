import { combineReducers } from 'redux';

import charactersReducer from './characters/characters.reducer';
import dialogReducer from './dialog/dialog.reducer';
import filmsReducer from './films/films.reducer';
import planetsReducer from './planets/planets.reducer';
import speciesReducer from './species/species.reducer';

export default combineReducers({
  characters: charactersReducer,
  films: filmsReducer,
  planets: planetsReducer,
  species: speciesReducer,
  dialog: dialogReducer,
});
