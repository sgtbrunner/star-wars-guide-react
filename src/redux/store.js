import { createStore, applyMiddleware } from 'redux';
import { persistStore } from 'redux-persist';
import { composeWithDevTools } from 'redux-devtools-extension';
import thunk from 'redux-thunk';

import rootReducer from './root-reducer';

const middlewares = [thunk];

export const store = createStore(rootReducer, composeWithDevTools(applyMiddleware(...middlewares)));

export const persistor = persistStore(store);

export default { store, persistor };
