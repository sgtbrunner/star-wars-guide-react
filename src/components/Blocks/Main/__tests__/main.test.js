import React from 'react';
import { render, screen } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import { Provider } from 'react-redux';
import { createStore } from 'redux';

import Main from '../main.component';

const loadedCharactersA = [
  {
    id: 0,
    name: 'Anakin Skywalker',
  },
  {
    id: 1,
    name: 'Obi-Wan Kenoby',
  },
  {
    id: 2,
    name: 'Yoda',
  },
];

const loadedCharactersB = [
  ...loadedCharactersA,
  {
    id: 3,
    name: 'Darth Vader',
  },
  {
    id: 4,
    name: 'Luke Skywalker',
  },
  {
    id: 5,
    name: 'Palpatine',
  },
];

const store = createStore(() => {});
const mountMainComponent = (characters) =>
  render(
    <Provider store={store}>
      <Main characters={characters} />
    </Provider>
  );

describe('Main component', () => {
  it('should match initial snapshot', () => {
    const wrapper = mountMainComponent(loadedCharactersB);

    expect(wrapper).toMatchSnapshot();
  });

  it(`should render all 6 (six) characters if nothing is input on searchbox`, () => {
    mountMainComponent(loadedCharactersB);

    // clear the searchbox
    const searchbox = screen.getByRole('searchbox');
    userEvent.clear(searchbox);

    // assert for existing cards
    const cards = screen.getAllByRole('button');
    expect(cards).toHaveLength(6);
    cards.forEach((card, index) => expect(card).toHaveTextContent(loadedCharactersB[index].name));
  });

  it(`should render 2 (two) characters only when names are filtered down to "sky",
    Anakin Skywalker and Luke Skywalker`, () => {
    mountMainComponent(loadedCharactersB);

    // type in "sky" on searchbox
    const searchbox = screen.getByRole('searchbox');
    userEvent.clear(searchbox);
    userEvent.type(searchbox, 'sky');

    // assert for existing cards
    const cards = screen.getAllByRole('button');
    expect(cards).toHaveLength(2);
    expect(cards[0]).toHaveTextContent(loadedCharactersB[0].name);
    expect(cards[1]).toHaveTextContent(loadedCharactersB[4].name);
  });

  it(`should render 1 (one) character only when names are filtered down to "darth",
    Darth Vader`, () => {
    mountMainComponent(loadedCharactersB);

    // type in "darth" on searchbox
    const searchbox = screen.getByRole('searchbox');
    userEvent.clear(searchbox);
    userEvent.type(searchbox, 'darth');

    // assert for existing cards
    const cards = screen.getAllByRole('button');
    expect(cards.length).toBe(1);
    expect(cards[0]).toHaveTextContent(loadedCharactersB[3].name);
  });

  it(`should render 'No characters found' alert if NO characters are found after
    filtering characters by 'gsrfgsegse' name`, () => {
    mountMainComponent(loadedCharactersB);

    // type in "gsrfgsegse" on searchbox
    const searchbox = screen.getByRole('searchbox');
    userEvent.clear(searchbox);
    userEvent.type(searchbox, 'gsrfgsegse');

    // assert that cards do not exist
    const cards = screen.queryAllByRole('button');
    expect(cards.length).toBe(0);

    // assert for "no characters found" alert
    const alert = screen.getByRole('alert');
    expect(alert).toBeInTheDocument();
    expect(alert).toHaveTextContent(/no characters found!/i);
  });
});
