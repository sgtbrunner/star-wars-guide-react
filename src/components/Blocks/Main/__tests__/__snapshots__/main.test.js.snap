// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Main component should match initial snapshot 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        id="main"
      >
        <div
          class="pa2 br3 searchbox"
        >
          <input
            class="f4 pa2 w-100 ba bw1 custom-field"
            placeholder="find your favorite Star Wars character..."
            type="search"
          />
        </div>
        <div
          class="cardlist"
        >
          <button
            class="card pointer"
            id="0"
            type="button"
          >
            <img
              alt="character-portrait"
              id="0"
              src="https://starwars-visualguide.com/assets/img/characters/1.jpg"
            />
            <div
              class="name"
              id="0"
            >
              Anakin Skywalker
            </div>
          </button>
          <button
            class="card pointer"
            id="1"
            type="button"
          >
            <img
              alt="character-portrait"
              id="1"
              src="https://starwars-visualguide.com/assets/img/characters/2.jpg"
            />
            <div
              class="name"
              id="1"
            >
              Obi-Wan Kenoby
            </div>
          </button>
          <button
            class="card pointer"
            id="2"
            type="button"
          >
            <img
              alt="character-portrait"
              id="2"
              src="https://starwars-visualguide.com/assets/img/characters/3.jpg"
            />
            <div
              class="name"
              id="2"
            >
              Yoda
            </div>
          </button>
          <button
            class="card pointer"
            id="3"
            type="button"
          >
            <img
              alt="character-portrait"
              id="3"
              src="https://starwars-visualguide.com/assets/img/characters/4.jpg"
            />
            <div
              class="name"
              id="3"
            >
              Darth Vader
            </div>
          </button>
          <button
            class="card pointer"
            id="4"
            type="button"
          >
            <img
              alt="character-portrait"
              id="4"
              src="https://starwars-visualguide.com/assets/img/characters/5.jpg"
            />
            <div
              class="name"
              id="4"
            >
              Luke Skywalker
            </div>
          </button>
          <button
            class="card pointer"
            id="5"
            type="button"
          >
            <img
              alt="character-portrait"
              id="5"
              src="https://starwars-visualguide.com/assets/img/characters/6.jpg"
            />
            <div
              class="name"
              id="5"
            >
              Palpatine
            </div>
          </button>
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      id="main"
    >
      <div
        class="pa2 br3 searchbox"
      >
        <input
          class="f4 pa2 w-100 ba bw1 custom-field"
          placeholder="find your favorite Star Wars character..."
          type="search"
        />
      </div>
      <div
        class="cardlist"
      >
        <button
          class="card pointer"
          id="0"
          type="button"
        >
          <img
            alt="character-portrait"
            id="0"
            src="https://starwars-visualguide.com/assets/img/characters/1.jpg"
          />
          <div
            class="name"
            id="0"
          >
            Anakin Skywalker
          </div>
        </button>
        <button
          class="card pointer"
          id="1"
          type="button"
        >
          <img
            alt="character-portrait"
            id="1"
            src="https://starwars-visualguide.com/assets/img/characters/2.jpg"
          />
          <div
            class="name"
            id="1"
          >
            Obi-Wan Kenoby
          </div>
        </button>
        <button
          class="card pointer"
          id="2"
          type="button"
        >
          <img
            alt="character-portrait"
            id="2"
            src="https://starwars-visualguide.com/assets/img/characters/3.jpg"
          />
          <div
            class="name"
            id="2"
          >
            Yoda
          </div>
        </button>
        <button
          class="card pointer"
          id="3"
          type="button"
        >
          <img
            alt="character-portrait"
            id="3"
            src="https://starwars-visualguide.com/assets/img/characters/4.jpg"
          />
          <div
            class="name"
            id="3"
          >
            Darth Vader
          </div>
        </button>
        <button
          class="card pointer"
          id="4"
          type="button"
        >
          <img
            alt="character-portrait"
            id="4"
            src="https://starwars-visualguide.com/assets/img/characters/5.jpg"
          />
          <div
            class="name"
            id="4"
          >
            Luke Skywalker
          </div>
        </button>
        <button
          class="card pointer"
          id="5"
          type="button"
        >
          <img
            alt="character-portrait"
            id="5"
            src="https://starwars-visualguide.com/assets/img/characters/6.jpg"
          />
          <div
            class="name"
            id="5"
          >
            Palpatine
          </div>
        </button>
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
